src/compiler/checker.ts

```diff
@@ -25203,6 +25203,48 @@ namespace ts {
             return literalType;
         }
 
+        /**
+         * Replaces all references to structurally equivalent object and array literal types in the given list
+         * with references to a single one of those types. The process is applied recursively to properties of
+         * object literals and elements of array literals.
+         */
+        function deduplicateObjectOrArrayLiteralTypes(types: Type[]) {
+            if (!some(types, isObjectOrArrayLiteralType)) {
+                return types;
+            }
+            const typeMap = new Map<string, Type>();
+            return sameMap(types, getInternedType);
+
+            function getInternedType(type: Type): Type {
+                if (type.flags & TypeFlags.Union) {
+                    const newTypes = sameMap((<UnionType>type).types, getInternedType);
+                    return newTypes !== (<UnionType>type).types ? getUnionType(newTypes) : type;
+                }
+                const key = getLiteralTypeKey(type);
+                return key ? typeMap.get(key) || (typeMap.set(key, type), type) : type;
+            }
+
+            function getLiteralTypeKey(type: Type): string | undefined {
+                const objectFlags = getObjectFlags(type);
+                if (objectFlags & ObjectFlags.ObjectLiteral) {
+                    const props = getPropertiesOfObjectType(type);

```

Comment:
I'm *pretty sure* we should sort this in some way, since I think our array ordering will respect the input order (our maps all respect insertion order for iteration at least) - so `{a, b}` and `{b, a}` are going to yield props in differing orders, which'll in turn make differing keys here, but are... essentially the same type? The same as far as we actually reason about anyway.

Reply:
Sorting wouldn't really make a difference because we'd sort by type ID and type IDs for literal types are assigned in order of appearance already. When types are structurally equivalent it's perfectly fine to just pick the first one.